src/Solvers.jl

```diff
@@ -18,45 +18,122 @@ using JuMP
+optimization variables.
+- `ub::Vector{Float64}`: Vector of size (n,) specifying the upper bounds for the
+optimization variables.
+
+# Returns
+- A tuple containing:
+  - `x`: The optimized values of the variables `x` (vector of size (n,)).
+  - `mse_opt`: The exponential of the objective function value at the optimum.
+"""
+function opt_solve(A::Matrix{Float64}, b::Vector{Float64}, x0::Vector{Float64},
+    lb::Vector{Float64}, ub::Vector{Float64})
 
     #mle = Model(optimizer_with_attributes(Ipopt.Optimizer, "print_level" => 0))
     mle = Model(NLopt.Optimizer)
     set_optimizer_attribute(mle, "algorithm", :LD_SLSQP) #LD_SLSQP
 
-    x0[x0 .< 0] .= 0
-    x0[x0 .> 1] .= 1
+    x0[x0.<0] .= 0
+    x0[x0.>1] .= 1
 
     @variable(mle, lb[i] <= x[i=1:length(x0)] <= ub[i])
     #@constraint(mle, sum(x) == 1)
-    for n in 1:length(x0) set_start_value(x[n], x0[n]) end
+    for n in 1:length(x0)
+        set_start_value(x[n], x0[n])
+    end
 
     #@NLexpression(mle, mse ,  sum( (b .- sum(x[i] .* A[:,i] for i in 1:length(x) )).^2    ) )
-    @expression(mle, mse ,  sum( (b .- sum(x[i] .* A[:,i] for i in 1:length(x) )).^2    ) )
-    
+    @expression(mle, mse, sum((b .- sum(x[i] .* A[:, i] for i in 1:length(x))) .^ 2))
+
     @NLobjective(mle, Min, mse)
 
     JuMP.optimize!(mle)
-    return value.(mle[:x]) , exp(objective_value(mle))
+    return value.(mle[:x]), exp(objective_value(mle))
 end
 
+"""
+    dolsq(A, b; method::String="default")
+
+Solve a least squares problem, finding the vector `x` that minimizes the residual
+||Ax - b||².
 
+# Arguments
+- `A`: Coefficient matrix of size (m, n).
+- `b`: Target vector of size (m,).
+- `method`: An optional keyword argument specifying the solver method.
+Available options are:
+    - `"default"`: Solve the system using the backslash operator (`\`).
+    - `"pinv"`: Use the pseudoinverse of `A` to compute the solution.
+    - `"qr"`: Use QR decomposition to solve the least squares problem.
+
+# Returns
+- `x`: Vector of size (n,) that minimizes the least squares residual.
+"""
 function dolsq(A, b; method::String="default")
     if method == "default"
         x = A \ b
     elseif method == "pinv"
-        x = pinv(A)*b
+        x = pinv(A) * b
     elseif method == "qr"
         #Q,R = qr(A)
         #x = inv(R)*Q'*b
         qrA = qr(A)
-        x = qrA\b
+        x = qrA \ b
     end
     return x
 end
 
+"""
+    bvls(A, b, x_lsq, lb, ub, tol::Float64, max_iter::Int64, verbose::Int64,
+         inverse_method::String)
+
+Solve a bounded value least squares problem, finding the vector `x` subject to lower (`lb`)
+and upper (`ub`) bounds, that minimizes the residual ||Ax - b||².
+
+- The function employs an iterative approach, adjusting the solution based on the specified
+  bounds.
+- Convergence is determined by whether KKT optimality condition (see
+[`compute_kkt_optimality`](@ref)) is within the specified tolerance.
```

Comment:
Does this need to be elaborated on or is this good? I.e. the deeper details on the two loops, free vs bounded variables, etc etc?

Reply:
This is enough!  Let's add a quick reference to this paper in the comments: https://www.stat.berkeley.edu/~stark/Preprints/bvls.pdf

Reply:
Done, added as a Notes section to the docstring.